File: src/main/java/com/example/demo/DemoApplication.java
package com.example.demo;

import lombok.Data;
import org.springframework.beans.factory.config.YamlPropertiesFactoryBean;
import org.springframework.boot.CommandLineRunner;
import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.boot.bind.PropertySourcesPropertyValues;
import org.springframework.boot.bind.RelaxedDataBinder;
import org.springframework.context.annotation.Bean;
import org.springframework.core.env.MutablePropertySources;
import org.springframework.core.env.PropertiesPropertySource;
import org.springframework.core.io.ClassPathResource;
import org.springframework.data.mongodb.core.MongoOperations;

import java.util.ArrayList;
import java.util.List;
import java.util.Properties;

@SpringBootApplication
public class DemoApplication {

    public static void main(String[] args) {
        SpringApplication.run(DemoApplication.class, args);
    }

    @Bean
    public CommandLineRunner initData(MongoOperations mongo) {
        return (String... args) -> {
            mongo.dropCollection(Book.class);
            mongo.createCollection(Book.class);
            getBooks().forEach(mongo::save);
        };
    }

    private List<Book> getBooks() {
        Properties yaml = loadBooksYaml();
        MutablePropertySources propertySources = new MutablePropertySources();
        propertySources.addFirst(new PropertiesPropertySource("books", yaml));
        Books books = new Books();
        RelaxedDataBinder binder = new RelaxedDataBinder(books);
        binder.bind(new PropertySourcesPropertyValues(propertySources));
        return books.getBooks();
    }

    private Properties loadBooksYaml() {
        YamlPropertiesFactoryBean properties = new YamlPropertiesFactoryBean();
        properties.setResources(new ClassPathResource("books.yml"));
        return properties.getObject();
    }

    @Data
    private static class Books {
        private List<Book> books = new ArrayList<>();
    }
}

File: src/main/kotlin/com/example/demo/DemoApplication.kt
package com.example.demo

import org.springframework.beans.factory.config.YamlPropertiesFactoryBean
import org.springframework.boot.CommandLineRunner
import org.springframework.boot.SpringApplication
import org.springframework.boot.autoconfigure.SpringBootApplication
import org.springframework.boot.context.properties.bind.Bindable
import org.springframework.boot.context.properties.bind.Binder
import org.springframework.boot.context.properties.source.MapConfigurationPropertySource
import org.springframework.context.annotation.Bean
import org.springframework.core.io.ClassPathResource
import org.springframework.data.mongodb.core.MongoOperations
import java.util.*

@SpringBootApplication
class DemoApplication {

    companion object {
        @JvmStatic
        fun main(args: Array<String>) {
            SpringApplication.run(DemoApplication::class.java, *args)
        }
    }

    private val books: List<Book>
        get() {
            val yaml = loadBooksYaml()
            val source = MapConfigurationPropertySource(yaml)
            return Binder(source).bind("books", Bindable.listOf(Book::class.java)).get()
        }

    @Bean
    fun initData(mongo: MongoOperations) = CommandLineRunner {
        mongo.dropCollection(Book::class.java)
        mongo.createCollection(Book::class.java)
        books.forEach { mongo.save(it) }
    }

    private fun loadBooksYaml(): Properties {
        val properties = YamlPropertiesFactoryBean()
        properties.setResources(ClassPathResource("books.yml"))
        return properties.`object`!!
    }
}
